### PR TITLE
Display Order of Card Browser corrected

### PR DESCRIPTION
**Pull request by tapish2000**
## Purpose / Description
Before, there was a bug - when we close the app after changing the display order in Card Browser, and then reopen it again, it was getting set to the default value, it was not getting saved in the DB. Adding just one command ie getCol().flush() solved this issue.

## Fixes
Issue #4199 

## Approach
Just added `getCol().flush()` in `CardBrowser.java` in the function `changeCardOrder()`. This saves the updated `sortType` field in the database.

## How Has This Been Tested?
1. From the homepage, go to Card Browser.
2. Go to Change Display Order from Options.
3. Select any order.
4. Close the app and remove it from Android History.
5. Reopen the app and go to Card Browser again.
6. The order selected will be seen.

## Learning (optional, can help others)
Just find out where the error is originating either by using debugging statements (`Timber`) or anything and then keep using `Ctrl + left mouse click` on the functions nearby to get to the root of the problem.

## Checklist

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
(https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by tapish2000** [comment]
Sorry for the late PR, I had been busy in some college work and I checked that this is corresponding to a `Priority-Low` Issue, so I didn't push myself harder.

**Review by david-allison-1 — CHANGES_REQUESTED** [review]
Thanks! Could you add a unit test to ensure that we don't get a regression

After the test, could you see if `setMod` works:

https://github.com/ankidroid/Anki-Android/blob/aa0c51bcae88ffbfa76224c27beed0c6d923dcd9/AnkiDroid/src/main/java/com/ichi2/libanki/Collection.java#L349-L355

**Comment by tapish2000** [comment]
@david-allison-1 Should I write a test to check if `setMod()` works (returns true) on changing the `sortType` field by the function `changeCardOrder()`?

**Comment by david-allison-1** [comment]
No, test that:

* if the value is set
* then the activity is closed (ideally via onPause/onDestory)
* then the activity is opened
* that the value change is persisted

Then we can change the implementation of the feature and the test will still pass

**Comment by tapish2000** [comment]
I wrote a test for the mentioned steps, what about `setMod` you said earlier, should I assert for that too in the test function or should I push it for you to verify first?

**Comment by david-allison-1** [comment]
> I wrote a test for the mentioned steps, what about `setMod` you said earlier, should I assert for that too in the test function or should I push it for you to verify first?

Nothing for me to verify: if `setMod` works, then the test should pass, if it doesn't, then the test should fail

**Comment by tapish2000** [comment]
In my recent commit, I am checking whether `sortType` is changing or not after the app is killed and restarted. I hope this is fine and if this works, I think `setMod` must have worked correctly.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 1]
```diff
@@ -527,6 +528,36 @@ public void checkSearchString() {
...
+        ensureCollectionLoadIsSynchronous();
+        ActivityController<CardBrowser> cardBrowserController = Robolectric.buildActivity(CardBrowser.class, new Intent())
+                .create().start();
+        cardBrowserController.resume().visible();
+        saveControllerForCleanup(cardBrowserController);
+        advanceRobolectricLooperWithSleep();
+        CardBrowser cardBrowser = cardBrowserController.get();
+
+        // Make sure card has default value in sortType field
+        assertThat("Initially Card Browser has order = noteFld", cardBrowser.getCol().getConf().get("sortType"), is("noteFld"));
```
Comment (david-allison-1) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:544]:
```suggestion
        assertThat("Initially Card Browser has order = noteFld", getCol().getConf().get("sortType"), is("noteFld"));
```

**Review by david-allison-1 — APPROVED** [review]
Looks great, thanks!

**Comment by Arthur-Milchior** [comment]
> Sorry for the late PR,

I don't understand.

I mean, The bug was opened in 2016 but I guess you were not considering yourself 5 years late.
I encoutered this bug often but never cared enough to look at it. I guess I just got used to things. I actually had this bug before I fist contributed to AnkiDroid, so I found it to be a natural part of the app I guess.


Anyway, thanks a lot. I entirely agree, it works! 

However, I will not approve it. But the nice side is that you made me discover that the problem I am going to mention already exists in the code, and if you agree, it would be great if you could fix it too. (Feel free to reject, then I'll open an issue or take care of it directly)

`flush` is an expensive operations. It takes json, transform into string and save it in DB. This is not something we actually need to do each time we want to modify the configuration. Instead, we should just note that the DB has changed, `getCol().getDb().setMod(true)` and let the `save` function write data later when needed.

This should have been true for all other call to `flush()` and the javadoc should indicates that actually, `flush` should only be used for testing (note that there is no `flush` anymore upstream)

What is even more important is that we should:
* check that all changes to the configurations are followed by this same call
* document on `getConf` what should be done to save the configuration changed

I'd ideally want to create a Configuration class to deal with it explicitly, to avoid future problem. But since we'll move to rust, it may be overkill

**Comment by tapish2000** [comment]
> I don't understand.
> 
> I mean, The bug was opened in 2016 but I guess you were not considering yourself 5 years late.
😂😂 I was considering the day when I was assigned this issue as my start point 😅.

I can try what you are suggesting once, if it's possible it will done in this PR only. 

Thanks!

**Comment by david-allison-1** [comment]
> This should have been true for all other call to `flush()` and the javadoc should indicates that actually, `flush` should only be used for testing (note that there is no `flush` anymore upstream)

The problem is Android allows the app to be killed at any time when minimised - (I haven't tested, but I don't think we don't call `save` when the app goes to the background), and after that Android can stop the process, discarding the change

**Comment by tapish2000** [comment]
I changed `flush()` to the better one `getCol().getDb().setMod(true)` so that `save` takes care of it. It worked. 

I see there are many other places where flush is used. What can be done for those?

**Comment by Arthur-Milchior** [comment]
> The problem is Android allows the app to be killed at any time when minimised - (I haven't tested, but I don't think we don't call `save` when the app goes to the background), and after that Android can stop the process, discarding the change

There are probably too many negations here.
Anyway, if you're right then that's a bug that should be corrected too. Because there are many things that wait and is not directly flushed, and that should probably be saved too.

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java [hunk 2]
```diff
@@ -527,6 +528,32 @@ public void checkSearchString() {
...
+                .create().start();
+        cardBrowserController.resume().visible();
+        saveControllerForCleanup(cardBrowserController);
+        advanceRobolectricLooperWithSleep();
+
+        // Make sure card has default value in sortType field
+        assertThat("Initially Card Browser has order = noteFld", getCol().getConf().get("sortType"), is("noteFld"));
+
+        // Change the display order of the card browser
+        getCol().getConf().put("sortType", "cardEase");
```
Comment (mikehardy) [inline AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java:546]: I love that there is a test, and I believe the *actual* change in the PR is good, that is I believe it fixes the problem. But I don't think the test is actually probing the change. The change was in `changeCardOrder` so it seems like the test should be:

```suggestion
        cardBrowserController.get().changeCardOrder(<something here>);
```

...probably with a change to that API to be public (so you may access it) and a `@VisibleForTesting(otherwise = VisibleForTesting.PROTECTED)` annotation.

Alternatively (maybe actually easier?) you can use findViewById I believe and operate the user interface by getting the actual user interface views and "clicking" them

then the assertion is not just that the sortType is the one you intend, but that the modification date has moved from it's previous value (which I guess you would have to access and save first, so you have it available for comparison.

In short: the test should save previous sort order and mod time so you can assert later, it should *actually exercise the method you changed - not just mutate the database directly* and then assert all the changes you made are reflected

Testing is sort of tedious that way but also it feels really good (or at least it does to me) when you just know that you nailed it.

**Review by mikehardy — CHANGES_REQUESTED** [review]
I like the change but I left a long testing request (apologies)

File: AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java (hunk 2 above)
Comment (tapish2000) [reply AnkiDroid/src/test/java/com/ichi2/anki/CardBrowserTest.java]: Thanks for asking me to improve this.

**Review by mikehardy — APPROVED** [review]
That's nice! I think the test is definitely exercising the change you made now - if changing the sortType and incrementing col.modTime is the fix (and we think it is because you tested it on device as well...) this unit test will make sure that's happening forever now. Thanks!

**Comment by mikehardy** [comment]
+1 on the clean commit history as well, rebase-merge is possible

**Comment by mikehardy** [comment]
Hi there!

Just a friendly notice that we try to process OpenCollective payments monthly - it's time for April 2021 submissions

If you are interested in compensation for this work, the process with details is here:

https://github.com/ankidroid/Anki-Android/wiki/OpenCollective-Payment-Process#how-to-get-paid

(I only post one comment per person to avoid spamming you, regardless of the number of PRs merged, but this note applies to all PRs merged in April)

Thanks!
